### PR TITLE
Reference attribute

### PR DESCRIPTION
**Pull request by tpluscode**

### CODE REVIEW

File: api/hydra/api.ttl [hunk 1]
```diff
@@ -22,7 +22,7 @@
     dataCube:Column ,
     api:Error ,
     api:ProjectPlaceholder ,
-    dataCube:ColumnAttribute ,
+    dataCube:ValueAttribute ,
```
Comment (tpluscode) [inline api/hydra/api.ttl]: Potential breaking change: I renamed the attribute as we called it "value attribute" wit hAdrian

File: api/hydra/api.ttl [hunk 2]
```diff
@@ -575,12 +576,12 @@ dataCube:ColumnAttribute rdfs:subclassOf dataCube:Attribute ;
     hydra:description "Cannot be used together with datatype"
   ] .
 
-api:AddAttributeMappedFromSingleColumn
-  a hydra:SupportedOperation, hydra-box:View ;
+_:PostValueAttribute
+  a hydra:SupportedOperation, hydra-box:View, api:AddValueAttribute ;
```
Comment (tpluscode) [inline api/hydra/api.ttl]: Potentially breaking, although I renamed the usage in UI code

```diff
-api:AddAttributeMappedFromSingleColumn
+api:AddValueAttribute
```

File: api/test/DimensionTable/AddDimensionAttribute.hydra [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+                Content-Type "text/turtle"
+
+                ```
+                @prefix dataCube: <https://rdf-cube-curation.described.at/> .
+                @prefix schema: <http://schema.org/> .
+                @prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
+
+                <> dataCube:referencedTable <project/reference-attribute-test/table/Stations> ;
+                   rdf:predicate <http://example.org/station> ;
+                   a dataCube:ReferenceAttribute ;
```
Comment (tpluscode) [inline api/test/DimensionTable/AddDimensionAttribute.hydra]: Note that explicit `rdf:type` is now needed as both attributes land as `POST /attributes` requests

**Comment by martinmaillard** [comment]
Is there a way to get a description of what an endpoint expects with Hydra?

File: api/hydra/api.ttl [hunk 4]
```diff
@@ -591,6 +592,23 @@ api:AddAttributeMappedFromSingleColumn
     )
   ] .
 
+_:PostReferenceAttribute
+  a hydra:SupportedOperation, hydra-box:View, api:AddReferenceAttribute ;
+  hydra:method "POST" ;
+  hydra:title "Creates a reference attribute" ;
+  hydra:description "Creates a table attribute referencing a dimension table" ;
+  hydra:expects dataCube:ReferenceAttribute ;
```
Comment (martinmaillard) [inline api/hydra/api.ttl]: I can't find a description of `ReferenceAttribute` in the file. Is it missing or not required?

Comment (tpluscode) [reply api/hydra/api.ttl]: Ah my mistake. Of course, will add it

**Comment by tpluscode** [comment]
>Is there a way to get a description of what an endpoint expects with Hydra?

As you find in `api.ttl` in code it's under `operation.supportedProperty.expects` which returns an `IClass` object

**Comment by martinmaillard** [comment]
Does this PR require running a migration on existing triplestores?

**Comment by tpluscode** [comment]
Ah good point. Yes, migration might be required

**Comment by martinmaillard** [comment]
What is the format that `columnMapping` expects? Probably something like:
```
[
  {'dataCube:sourceColumn': <id>, 'dataCube:referencedColumn': <id>},
   ...
]
```
The description `Identifier column in source table` mentions a `column`. Are we talking about a CSV column, or another attribute?

**Comment by tpluscode** [comment]
Meaning the API identifier URI.

Have a look at the example payload in e2e test cases:
https://github.com/zazuko/data-cube-curation/blob/c88b76438cde806cf1b74deb36af166c5335294d/api/test/DimensionTable/AddDimensionAttribute.hydra#L94-L107

**Comment by martinmaillard** [comment]
Ok thanks. Then it would be great if, for each `Table`, I could the list of columns extracted from the `identifierTemplate`.

**Comment by tpluscode** [comment]
Indeed. This came to my mind and I wasn't sure if we actually need it or not. Good point.

Would you agree to handle this in a separate PR?

**Comment by martinmaillard** [comment]
@tpluscode I think Alcaeus has trouble understanding what's is provided in `columnMapping`. It gives me an array of `OperationFinderMixinClass`

**Comment by tpluscode** [comment]
That mixing class is the implementation of the `findOperationsDeep` and other such methods. The runtime class name is irrelevant though. The objects should all implement `HydraResource` as usual.

**Comment by martinmaillard** [comment]
Mmh... ok, then I guess it's not the issue. But there's not much in there: 
Screenshot 2019-12-10 at 21 23 54

**Comment by martinmaillard** [comment]
@tpluscode Isn't it missing the `referencedColumn` and `sourceColumn`?

**Comment by tpluscode** [comment]
oof, sorry I forgot about this.

Is there a branch I can run and investigate?

**Comment by martinmaillard** [comment]
I have a branch here: https://github.com/zazuko/data-cube-curation/tree/ui-ref-attribute

It's still very WIP though...

**Pull request merged**